core/src/main/scala/org/apache/spark/sql/SnappyParser.scala

```diff
@@ -348,6 +348,42 @@ class SnappyParser(session: SnappySession)
       (d, s.asInstanceOf[Option[Duration]]))
   }
 
+  private def extractGroupingSet(
+      child: LogicalPlan,
+      aggregations: Seq[NamedExpression],
+      groupByExprs: Seq[Expression],
+      groupingSets: Seq[Seq[Expression]]): GroupingSets = {
+    val keyMap = groupByExprs.zipWithIndex.toMap
+    val bitmasks: Seq[Int] = groupingSets.map(set => set.foldLeft(0)((bitmap, col) => {
+      require(keyMap.contains(col), s"$col doesn't show up in the GROUP BY list")
+      bitmap | 1 << keyMap(col)
```

Comment:
I had referred to HiveQl.scala from 1.6 code which was doing (bitmap | 1 << keyMap(col)).  

https://github.com/apache/spark/blob/branch-1.6/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala#L488

The new 2.0 code in AstBuilder.withAggregation is setting the bits differently as you mentioned (did not find any bug filed for it). Have now changed the make it similar as 2.0 as this one appears to be correct. Let me know what you think.